### PR TITLE
Overcloud network IPAM CRDs

### PR DESCRIPTION
**Pull request by dprince**
Includes 2 new CRDs for Overcloudnet and Overcloudipset. Using these we'll be able to model both the ctlplane and network isolation. The CRDs will automatically generate a ConfigMap named predictable-ips-heat-env which contains generated Heat env files for the deployed-server-map (required for TripleO's deployed servers) and additional predictable Ips env files for other networks.

### CODE REVIEW

File: controllers/suite_test.go [hunk 1]
```diff
@@ -60,6 +60,12 @@ var _ = BeforeSuite(func(done Done) {
 	err = ospdirectorv1beta1.AddToScheme(scheme.Scheme)
 	Expect(err).NotTo(HaveOccurred())
 
+err = ospdirectorv1beta1.AddToScheme(scheme.Scheme)
+Expect(err).NotTo(HaveOccurred())
+
+err = ospdirectorv1beta1.AddToScheme(scheme.Scheme)
+Expect(err).NotTo(HaveOccurred())
+
```
Comment (stuggi) [inline controllers/suite_test.go:68]: not needed, that got added by the sdk

File: templates/overcloudipset/config/ips-from-pool.yaml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+# Environment file for pre-assigned IPs
+
+parameter_defaults:
+{{.PredictableIps}}
```
Comment (stuggi) [inline templates/overcloudipset/config/ips-from-pool.yaml:4]: as mentioned above we can use {{ .PredictableIps | nindent X }}

File: api/v1beta1/overcloudnet_types.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+// OvercloudNetSpec defines the desired state of OvercloudNet
+type OvercloudNetSpec struct {
+
+	// +kubebuilder:validation:Required
+	// Cidr the cidr to use for this network
+	Cidr string `json:"cidr"`
+
+	// +kubebuilder:validation:Required
+	// AllocationStart a set of IPs that are reserved and will not be assigned
+	AllocationStart string `json:"allocationStart,omitempty"`
```
Comment (abays) [inline api/v1beta1/overcloudnet_types.go:39]: Just curious here...are the `// +kubebuilder:validation:Required` and `omitempty` competing with one another?  My understanding was that `omitempty` made a field optional.

Comment (dprince) [reply api/v1beta1/overcloudnet_types.go:39]: yeah, let me revisit those validations. When I initially created the APIs I was thinking those would be optional but after implementing it I think we probably do just want to require them.

File: controllers/overcloudipset_controller.go [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+
+// deployedServerPortMap -
+func deployedServerPortMap(r map[string]string, c string) string {
+	var b bytes.Buffer
+	for hostname, ipaddr := range r {
+		b.WriteString(fmt.Sprintf("    %s-ctlplane:\n", hostname))
+		b.WriteString(fmt.Sprintf("      fixed_ips:\n"))
+		b.WriteString(fmt.Sprintf("        - ip_address: %s\n", ipaddr))
+		b.WriteString(fmt.Sprintf("      subnets:\n"))
+		b.WriteString(fmt.Sprintf("        - cidr: %s\n", c))
```
[the remark this answers is not in the document]
Comment (dprince) [reply controllers/overcloudipset_controller.go:238]: @stuggi THanks! I created this github issue to capture this idea: https://github.com/openstack-k8s-operators/osp-director-operator/issues/51

File: api/v1beta1/overcloudnet_types.go (hunk 3 above)
Comment (dprince) [reply api/v1beta1/overcloudnet_types.go]: @abays fixed in the latest push

File: controllers/suite_test.go (hunk 1 above)
Comment (dprince) [reply controllers/suite_test.go]: @stuggi removed in latest push

**Review by abays — APPROVED** [review]

**Review by stuggi — APPROVED** [review]

**Pull request merged**